dotcom-rendering/src/model/groupTrailsByDates.ts

```diff
@@ -0,0 +1,207 @@
+		const existingYear = trailsByYear.find(
+			({ year }) => year === date.getFullYear(),
+		);
+		if (existingYear) {
+			existingYear.trails.push({ trail, date });
+		} else {
+			trailsByYear.push({
+				year: date.getFullYear(),
+				trails: [{ trail, date }],
+			});
+		}
+	});
+
+	return trailsByYear;
+};
+
+/**
+ * Takes a set of trails for a given year and returns them grouped by month of publication
+ */
+const groupTrailsByMonth = (trails: TrailAndDate[], year: number) => {
+	const trailsByMonth: Array<{
+		year: number;
+		month: number;
+		trails: TrailAndDate[];
+	}> = [];
+
+	trails.forEach(({ trail, date }) => {
+		const existingMonth = trailsByMonth.find(
+			({ month }) => month === date.getMonth(),
+		);
+		if (existingMonth) {
+			existingMonth.trails.push({ trail, date });
+		} else {
+			trailsByMonth.push({
+				year,
+				month: date.getMonth(),
+				trails: [{ trail, date }],
+			});
+		}
+	});
+
+	return trailsByMonth;
+};
+
+/**
+ * Takes a set of trails for a given year & month and returns them grouped by day of publication
+ */
+const groupTrailsByDay = (
+	trails: TrailAndDate[],
+	year: number,
+	month: number,
+) => {
+	const trailsByDay: Array<{
+		year: number;
+		month: number;
+		day: number;
+		trails: TrailAndDate[];
+	}> = [];
+
+	trails.forEach(({ trail, date }) => {
+		const existingMonth = trailsByDay.find(
+			({ day }) => day === date.getUTCDate(),
+		);
+		if (existingMonth) {
+			existingMonth.trails.push({ trail, date });
+		} else {
+			trailsByDay.push({
+				year,
+				month,
+				day: date.getUTCDate(),
+				trails: [{ trail, date }],
+			});
+		}
+	});
+
+	return trailsByDay;
+};
+
+/**
+ * This function takes a set of trails (DCRFrontCard[]), and groups them by day, month and sometimes day.
+ *
+ * All trails will be grouped by year & month, regardless of their frequency, we then decide whether or not to
+ * group them by day by looking at the mean rate of occurrence per-day, when this goes above
+ * a pre-determined value, the entire month is 'popped out' & trails are grouped by day instead of month.
+ *
+ * You can force the 'pop out' behavior with the 'forceDay' parameter.
+ *
+ * > Note: It's important to keep in mind that you won't get only get one of grouping by day, month & year and
+ * just month and year. Pop-outs happen on a per-month basis, so it's likely that some groups will be just year & month
+ * and others grouped by day, month & year.
+ *
+ * @param trails Trails to be grouped
+ * @param forceDay Optionally force the trails to be grouped by day, regardless of their frequency rate
+ * @returns Trails grouped by either month or day
+ */
+export const groupTrailsByDates = (
+	trails: DCRFrontCard[],
+	forceDay = false,
+): GroupedTrails[] => {
+	// Creates a helper type with easy access to the date of publication as a Date class
+	const trailsAndDates = trails.map((trail) => ({
+		trail,
+		date: trail.webPublicationDate
+			? new Date(trail.webPublicationDate)
+			: new Date(),
```

Comment:
Is this not the same as:

```suggestion
		date: new Date(trail.webPublicationDate),
```

Reply:
While it literally is, TS won't allow it, there's a overload with no parameters, and one with a `value: string` but they'd need one with `value: string | undefined` in order for that to work!